lib/AST/GenericEnvironment.cpp

```diff
@@ -53,6 +53,9 @@ GenericEnvironment::GenericEnvironment(
     // FIXME: If multiple generic parameters map to the same archetype,
     // the reverse mapping order is not deterministic.
   }
+
+  // Make sure this generic environment gets destroyed.
+  signature->getASTContext().addDestructorCleanup(*this);
```

Comment:
Eventually, yes. The problem is that `Type::subst()`, which is used a *lot* with these mappings in `GenericEnvironment`, requires a`TypeSubstitutionMap`. So we'd be unpacking the mapping a whole lot of times.

A better approach would be to teach `Type::subst()` and related not to use a `DenseMap` as their way of providing mappings. That'll take a bit more work, but is the right thing to do long-term.